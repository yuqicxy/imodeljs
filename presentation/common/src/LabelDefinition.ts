/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Hierarchies
 */

/**
 * Data structure that describes raw composite label value.
 * @public
 */
export interface LabelCompositeValue {
  separator: string;
  values: LabelDefinition[];
}

/** @internal */
export interface LabelCompositeValueJSON {
  separator: string;
  values: LabelDefinitionJSON[];
}

/**
 * @public
 */
export namespace LabelCompositeValue {
  /** @internal */
  export function toJSON(compositeValue: LabelCompositeValue): LabelCompositeValueJSON {
    return { ...compositeValue, values: compositeValue.values.map(LabelDefinition.toJSON) };
  }

  /** @internal */
  export function fromJSON(json: LabelCompositeValueJSON): LabelCompositeValue {
    return Object.assign({}, json, {
      values: json.values.map(LabelDefinition.fromJSON),
    });
  }
}

/**
 * Type definition of label raw value.
 * @public
 */
export type LabelRawValue = string | number | boolean | LabelCompositeValue;
/** @internal */
export type LabelRawValueJSON = string | number | boolean | LabelCompositeValueJSON;

/**
 * Data structure that describes label definition.
 * @public
 */
export interface LabelDefinition {
  /** Display value of label */
  displayValue: string;
  /** Raw value of label */
  rawValue: LabelRawValue;
  /** Type name of raw value */
  typeName: string;
}

/** @internal */
export interface LabelDefinitionJSON {
  displayValue: string;
  rawValue: LabelRawValueJSON;
  typeName: string;
}

/** @public */
export namespace LabelDefinition {
  /**
   * Serialize given label definition to JSON.
   * @internal
   */
  export function toJSON(labelDefinition: LabelDefinition): LabelDefinitionJSON {
    return {
      displayValue: labelDefinition.displayValue,
      typeName: labelDefinition.typeName,
      rawValue: isCompositeDefinition(labelDefinition) ? LabelCompositeValue.toJSON(labelDefinition.rawValue) : labelDefinition.rawValue,
    };
  }

  /**
   * Deserialize label definition from JSON
   * @param json JSON or JSON serialized to string to deserialize from
   * @returns Deserialized label definition
   *
   * @internal
   */
  export function fromJSON(json: LabelDefinitionJSON | string): LabelDefinition {
    if (typeof json === "string")
      return JSON.parse(json, reviver);
    return Object.assign({}, json, {
      rawValue: isCompositeDefinition(json) ? LabelCompositeValue.fromJSON(json.rawValue) : json.rawValue,
    });
  }

  /**
   * Reviver function that can be used as a second argument for
   * `JSON.parse` method when parsing [[LabelDefinition]] objects.
   *
   * @internal
   */
  export function reviver(key: string, value: any): any {
    return key === "" ? fromJSON(value) : value;
  }

  /** @internal */
  export function isCompositeDefinition(def: LabelDefinition): def is LabelDefinition & { rawValue: LabelCompositeValue } {
    return def.typeName === "composite";
  }

  /** @internal */
  export function fromLabelString(label: string): LabelDefinitionJSON {
    return {
      displayValue: label,
      rawValue: label,
      typeName: "string",
    };
  }
}
