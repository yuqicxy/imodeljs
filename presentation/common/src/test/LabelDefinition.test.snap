// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelCompositeValue fromJSON creates valid LabelCompositeValue from JSON 1`] = `
Object {
  "separator": "0",
  "values": Array [
    Object {
      "displayValue": "withdrawal",
      "rawValue": "withdrawal",
      "typeName": "string",
    },
    Object {
      "displayValue": "Practical",
      "rawValue": "Shoal",
      "typeName": "string",
    },
  ],
}
`;

exports[`LabelCompositeValue toJSON serializes LabelCompositeValue 1`] = `
Object {
  "separator": "5",
  "values": Array [
    Object {
      "displayValue": "B2B",
      "rawValue": "New Jersey",
      "typeName": "string",
    },
    Object {
      "displayValue": "Checking Account",
      "rawValue": "Knoll",
      "typeName": "string",
    },
  ],
}
`;

exports[`LabelDefinition fromJSON creates valid LabelDefinition from JSON 1`] = `
Object {
  "displayValue": "Sleek Concrete Bike",
  "rawValue": "deposit",
  "typeName": "string",
}
`;

exports[`LabelDefinition fromJSON creates valid LabelDefinition from serialized JSON 1`] = `
Object {
  "displayValue": "deposit",
  "rawValue": "Computer",
  "typeName": "string",
}
`;

exports[`LabelDefinition fromJSON creates valid LabelDefinition with composite value from JSON 1`] = `
Object {
  "displayValue": "Savings Account",
  "rawValue": Object {
    "separator": "t",
    "values": Array [
      Object {
        "displayValue": "Pines",
        "rawValue": "Savings Account",
        "typeName": "string",
      },
      Object {
        "displayValue": "implementation",
        "rawValue": "Auto Loan Account",
        "typeName": "string",
      },
    ],
  },
  "typeName": "composite",
}
`;

exports[`LabelDefinition fromJSON creates valid LabelDefinition with composite value from serialized JSON 1`] = `
Object {
  "displayValue": "Switchable",
  "rawValue": Object {
    "separator": "f",
    "values": Array [
      Object {
        "displayValue": "South Carolina",
        "rawValue": "collaboration",
        "typeName": "string",
      },
      Object {
        "displayValue": "Dominican Peso",
        "rawValue": "budgetary management",
        "typeName": "string",
      },
    ],
  },
  "typeName": "composite",
}
`;

exports[`LabelDefinition fromJSON creates valid Node from serialized JSON 1`] = `
Object {
  "displayValue": "Facilitator",
  "rawValue": "Rhode Island",
  "typeName": "string",
}
`;

exports[`LabelDefinition fromLabelString creates valid LabelDefinition from label string 1`] = `
Object {
  "displayValue": "Chief",
  "rawValue": "Chief",
  "typeName": "string",
}
`;

exports[`LabelDefinition toJSON serializes LabelDefinition 1`] = `
Object {
  "displayValue": "Gibraltar Pound",
  "rawValue": "Computer",
  "typeName": "string",
}
`;

exports[`LabelDefinition toJSON serializes LabelDefinition with composite value 1`] = `
Object {
  "displayValue": "navigate",
  "rawValue": Object {
    "separator": "r",
    "values": Array [
      Object {
        "displayValue": "intangible",
        "rawValue": "Cotton",
        "typeName": "string",
      },
      Object {
        "displayValue": "Tunnel",
        "rawValue": "structure",
        "typeName": "string",
      },
    ],
  },
  "typeName": "composite",
}
`;
